File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        #utilize 2 pointer approach starting points being left and right pointers
        left = 0
        right = len(height) - 1

        #calculate the area of each instance of our pointers, and then overwrite the max area each time we iterate over the bar heights and move our pointers
        max_area = 0

        while left < right:
            width = right - left
            h = min(height[left], height[right])
            area = width * h
            max_area = max(area, max_area)
            # print(f'left = {left} , right = {right}, width = {width}, area = {area}')
            #increment pointers for whichever is smaller to move towards the center
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
            
        return max_area

File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        // two pointer approach
        int l = 0; // left
        int r = height.size() - 1; // right
        int w; // width
        int h; // height
        int maxA = 0; // store the max at any given point
        while(l < r){
            w = r - l;
            h = min(height[l],height[r]);
            maxA = max(maxA, h * w);
            //cout << "h : " << h << ", w: " << w << " ";
            //cout << "area: " << h * w << " , maxA: " << maxA << endl;
            if(height[l] < height[r]){
                l++;
            } else if(height[r] < height[l]){
                r--;
            } else {
                l++;
            }
        }
        return maxA;
    }
};